app/Http/Middleware/ContentType.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+namespace App\Http\Middleware;
+
+use App\Exceptions\Internal\LycheeInvalidArgumentException;
+use App\Exceptions\UnexpectedContentType;
+use Illuminate\Http\Request;
+
+/**
+ * Class ContentType.
+ *
+ * This middleware ensures that the request has sent the correct content type
+ * The supported parameters are `:json` and `:multipart'.
+ */
+class ContentType
+{
+	public const JSON = 'json';
+	public const MULTIPART = 'multipart';
+
+	/**
+	 * Handles the incoming request.
+	 *
+	 * @param Request  $request     the incoming request to serve
+	 * @param \Closure $next        the next operation to be applied to the
+	 *                              request
+	 * @param string   $contentType the content type of the request's body;
+	 *                              either {@link ContentType::JSON} or
+	 *                              {@link ContentType::MULTIPART}
+	 *
+	 * @return mixed
+	 *
+	 * @throws UnexpectedContentType
+	 * @throws LycheeInvalidArgumentException
+	 */
+	public function handle(Request $request, \Closure $next, string $contentType): mixed
+	{
+		if ($contentType === self::JSON) {
+			if (!$request->isJson()) {
+				throw new UnexpectedContentType(self::JSON);
+			}
+		} elseif ($contentType === self::MULTIPART) {
+			if ($request->getContentType() !== 'form') {
+				throw new UnexpectedContentType(self::MULTIPART);
+			}
+		} else {
+			throw new LycheeInvalidArgumentException('$contentType must either be "' . self::JSON . '" or "' . self::MULTIPART . '"');
+		}
+
+		return $next($request);
```

Comment:
It seems that at least some error conditions (server crashes?) result in a `Syntax error` exception in this line rather than a helpful response pinpointing the actual problem. We need to fix it somehow because it makes debugging a real pain...

It's easily reproducible as I type this by setting `has_exiftool` to `0` and uploading any random picture.

Reply:
> It seems that at least some error conditions (server crashes?) result in a `Syntax error` exception in this line rather than a helpful response pinpointing the actual problem.

Yes, indeed these are server crashes.

A complete backtrace of a typical Laraval crash has hundreds of lines with dozens of unreadable anonymous closures in between which I personally find very difficult to read. Usually, I want to see the method of our own code which was involved in the problem.


So I thought it would be a clever idea not to print the whole backtrace into the log, but only print the _most inner method_ of _our_ code per exception.

See here:

https://github.com/LycheeOrg/Lychee/blob/cedfe3be5c9cc9abe4bd5fe09e1508e53ee7badb/app/Exceptions/Handler.php#L201

However, the problem which you observe happens when Lychee crashes after the middleware has been executed, but before the actual controller method has been reached. In this case the most inner of our own methods is one of our middleware methods.

Normally, the controller method should always be reached at least and if not this most likely indicates a severe error like a missing route, a unloadable class file, i.e. something really bad. I did not expect this to happen often.

But now I see that my initial idea might not have been so clever at all. 

> We need to fix it somehow because it makes debugging a real pain...

Right, but I wonder what would be the best approach?

The easiest solution would be not to try to be clever and not shorten the backtrace, but simply print the whole backtrace to the log. However, I fear that this will needlessly clutter the log for ordinary exceptions of "ordinary" use cases, i.e. after the controller method has been reached.

Another solution would be to let it depend on the environment. For "development" environments the whole backtrace is logged, for "production" environments a shortened backtrace is logged.

Or we could try to be even more clever and try to determine if the exception occurred before or after a controller method has been reached and then log different variants of backtraces. But I don't like the sound of that idea.

Reply:
> It's easily reproducible as I type this by setting has_exiftool to 0 and uploading any random picture.

Strangely, it does not happen for me:

![Screenshot_20220416_143424](https://user-images.githubusercontent.com/7428426/163675238-cfe40498-df69-42bf-b2b2-bc9c44a765da.png)

![Screenshot_20220416_143515](https://user-images.githubusercontent.com/7428426/163675243-6ed771f5-8c81-42cf-a9ab-67a2ea18d927.png)

![Screenshot_20220416_143533](https://user-images.githubusercontent.com/7428426/163675247-b1be1386-0951-4612-9694-7447d2499a0b.png)

Reply:
Even though, I could not trigger the error in the same way as @kamil4, I triggered a similar error by deliberately introducing a crash inside the Laravel framework.

I fixed the problem here: https://github.com/LycheeOrg/Lychee/pull/1217/commits/b3daac3bb63707c3fca761f6a3c03bbc9b01bf22

I went a little bit over the edge and decided to log up to two frames per exception. @kamil4: What do you think about it?

Reply:
Now, an early crash is supposed to look like this.

![Screenshot_20220416_202636](https://user-images.githubusercontent.com/7428426/163687073-f8e456be-f109-4cb5-b212-2ec7ce5dd98e.png)

You see the actual method which has thrown the exception (e.g. `ValidatePostSize`) and the most inner method of our own code (e.g. `FixStatusCode`). The latter won't be helpful for error cases like these, when the exception occurred before the controller method has been reached, but it doesn't harm either. The message `just a test` was the message of the exception I inserted deliberately into the code in order to provoke the problem.

Reply:
Thanks for working on this. I'm still struggling to debug the issue I'm seeing though. Here's what I see in the logs now:
```
2022-04-21 19:50:15 UTC -- error   -- Illuminate\Http\JsonResponse::setData -- 88 -- Syntax error
2022-04-21 19:50:15 UTC -- error   -- App\Http\Middleware\ContentType::handle -- 49 -- Syntax error; caused by
```
I don't think I'm experiencing a crash. Rather, the serialization of newly added `Photo` model is failing. That happens for me if `has_exifool = 0`, even for images without EXIF data. The same photo model object persists successfully into the DB, and I verified that there's no weirdness there. So why is it failing here while being turned into json?

Reply:
WTF? `JSON::setData` is failing?! That's extremely weird. I assume it might be a `null` pointer issue, but that is just a guess. And why it is failing for you, while I cannot reproduce the problem.

Could you look into the debug panel of your browser and post the response? The Lychee logs only contain the two most relevant entries, because I didn't want to clutter the logs with a full backtrace. But the JSON response should contain the full backtrace.

Wait, you don't get a JSON response, because `JSON::setData` fails, right? 😂

Reply:
The only positive effect of the new logging is that I can now tell you exactly that it fails in line 88, i.e. here https://github.com/laravel/framework/blob/f0d10fd788a917e005c6dcbf9ade1b21bc685c73/src/Illuminate/Http/JsonResponse.php#L88

So you have a proper step debugger which hooks into the PHP process when an exception is thrown? I am really curious what `$data` is.

Moreover, the error message is strange. `json_last_error` should return the last error message of JSON extension, but "syntax error" doesn't make any sense, if the last action was _encoding_. A syntax error makes only sense for _decoding_.

Reply:
I do not have a PHP step debugger -- I'm forced to add `Logs::notice` all over the place (because the debugbar doesn't work for uploads). `$data` is just fine, BTW.

So I figured it out, to a degree. Not the root cause, but why it fails here. It fails because `setData` is coded to assume that no exceptions are being thrown, so after invoking `toJson` on the model it calls `json_last_error()` to check if it succeeded. But our custom `toJson` forces the use of exceptions: https://github.com/LycheeOrg/Lychee/blob/324188c65316d49e4b2fcd79366922251c1a08cd/app/Models/Extensions/ThrowsConsistentExceptions.php#L134
So `json_last_error()` returns a stale error code from who-knows-where-and-when... As you pointed out, probably from some `json_decode`, given the `Syntax error`. I'm guessing that it works with ExifTool because its php-exif adapter invokes `json_decode` so the error gets cleared.

I was looking for a way to clear `json_last_error()` but I couldn't find one. I suppose we could do something crazy like invoking `json_encode` on an empty string. Still, fragile and nasty. I guess the correct fix is to provide a custom `setData` that works with exceptions? I'll let you figure it out...

Reply:
I cannot follow your explanation.

> It fails because `setData` is coded to assume that no exceptions are being thrown, so after invoking `toJson` on the model it calls `json_last_error()` to check if it succeeded. But our custom `toJson` forces the use of exceptions.

This doesn't make sense to me. If `json_last_error` had already reported an error in our `toJson`, then our `toJson` method would have thrown an exception itself and the failing line in `setData` would never have been reached. The fact that line 88 in `toData` is executed, means that `toJson` has not thrown an exception which in turn means that `json_last_error` was just fine back there. Of course only, if it is our `toJson` which is called at all.

So, there are only three options:

 1. `$data` is one of our models, our `toJson` is called, does not throw an exception which means `json_last_error` is fine and then suddenly an error slips in between line 78 and 88 of `toData`.
 2. `$data` is an instance of `Jsomable`, but not one of our models, `toJson` of some other object is called and this one sets `json_last_error`
 3. `$data` is not an instance of `Jsonable` at all and `json_encode`  in line 80, 82 or 84 is called.

Option 1 is impossible. So option 2 and 3 remain.


> `$data` is just fine, BTW.

What exactly is `$data`? It cannot be one of our models.


> So `json_last_error()` returns a stale error code from who-knows-where-and-when... As you pointed out, probably from some `json_decode`, given the `Syntax error`.

And the question is from where? Any successful `json_...` method resets the error. Given the method `toData` this can only originate from one of the `json_...` methods in there or from a `toJson`, but not ours.

> I'm guessing that it works with ExifTool because its php-exif adapter invokes `json_decode` so the error gets cleared.

> I was looking for a way to clear `json_last_error()` but I couldn't find one. I suppose we could do something crazy like invoking `json_encode` on an empty string.

Not necessary. If any of the `json_...` methods in `toData` was successful, this would clear that error.


Anyway, I really need to reproduce this error myself. Could you post me your exact setup? Webserver, PHP version, Lychee configuration, etc. This is crazy.

Or maybe just before line https://github.com/laravel/framework/blob/6fea6c271086e2ad017a6078e1ee457a2b8c398e/src/Illuminate/Http/JsonResponse.php#L88, could you dump the backtrace to console via [`debug_print_backtrace`](https://www.php.net/manual/en/function.debug-print-backtrace.php) and then `die`?

Reply:
Here's the output of `debug_print_backtrace` you wanted:
```
#0  Illuminate\Http\JsonResponse->setData() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/symfony/http-foundation/JsonResponse.php:54]
#1  Symfony\Component\HttpFoundation\JsonResponse->__construct() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Http/JsonResponse.php:32]
#2  Illuminate\Http\JsonResponse->__construct() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/Router.php:808]
#3  Illuminate\Routing\Router::toResponse() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/Router.php:789]
#4  Illuminate\Routing\Router->prepareResponse() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/Router.php:721]
#5  Illuminate\Routing\Router->Illuminate\Routing\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:128]
#6  Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/app/Http/Middleware/ContentType.php:49]
#7  App\Http\Middleware\ContentType->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#8  Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php:50]
#9  Illuminate\Routing\Middleware\SubstituteBindings->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#10 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php:78]
#11 Illuminate\Foundation\Http\Middleware\VerifyCsrfToken->handle() called at [/home/kamil/public_html/tmp/Lychee-json/app/Http/Middleware/VerifyCsrfToken.php:61]
#12 App\Http\Middleware\VerifyCsrfToken->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#13 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Session/Middleware/AuthenticateSession.php:39]
#14 Illuminate\Session\Middleware\AuthenticateSession->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#15 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php:49]
#16 Illuminate\View\Middleware\ShareErrorsFromSession->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#17 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php:121]
#18 Illuminate\Session\Middleware\StartSession->handleStatefulRequest() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php:64]
#19 Illuminate\Session\Middleware\StartSession->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#20 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php:37]
#21 Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#22 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php:67]
#23 Illuminate\Cookie\Middleware\EncryptCookies->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#24 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/app/Http/Middleware/AcceptContentType.php:61]
#25 App\Http\Middleware\AcceptContentType->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#26 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:103]
#27 Illuminate\Pipeline\Pipeline->then() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/Router.php:723]
#28 Illuminate\Routing\Router->runRouteWithinStack() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/Router.php:698]
#29 Illuminate\Routing\Router->runRoute() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/Router.php:662]
#30 Illuminate\Routing\Router->dispatchToRoute() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/Router.php:651]
#31 Illuminate\Routing\Router->dispatch() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:167]
#32 Illuminate\Foundation\Http\Kernel->Illuminate\Foundation\Http\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:128]
#33 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/livewire/livewire/src/DisableBrowserCache.php:19]
#34 Livewire\DisableBrowserCache->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#35 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/barryvdh/laravel-debugbar/src/Middleware/InjectDebugbar.php:67]
#36 Barryvdh\Debugbar\Middleware\InjectDebugbar->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#37 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/bepsvpt/secure-headers/src/SecureHeadersMiddleware.php:21]
#38 Bepsvpt\SecureHeaders\SecureHeadersMiddleware->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#39 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php:21]
#40 Illuminate\Foundation\Http\Middleware\TransformsRequest->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ConvertEmptyStringsToNull.php:31]
#41 Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#42 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php:21]
#43 Illuminate\Foundation\Http\Middleware\TransformsRequest->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TrimStrings.php:40]
#44 Illuminate\Foundation\Http\Middleware\TrimStrings->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#45 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php:27]
#46 Illuminate\Foundation\Http\Middleware\ValidatePostSize->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#47 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php:86]
#48 Illuminate\Foundation\Http\Middleware\PreventRequestsDuringMaintenance->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#49 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/fideloper/proxy/src/TrustProxies.php:57]
#50 Fideloper\Proxy\TrustProxies->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#51 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/app/Http/Middleware/FixStatusCode.php:31]
#52 App\Http\Middleware\FixStatusCode->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#53 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/itsgoingd/clockwork/Clockwork/Support/Laravel/ClockworkMiddleware.php:24]
#54 Clockwork\Support\Laravel\ClockworkMiddleware->handle() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#55 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:103]
#56 Illuminate\Pipeline\Pipeline->then() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:142]
#57 Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter() called at [/home/kamil/public_html/tmp/Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:111]
#58 Illuminate\Foundation\Http\Kernel->handle() called at [/home/kamil/public_html/tmp/Lychee-json/public/index.php:74]
```
Now, I owe you an apology for giving you an abbreviated version of my findings in my previous message, which sent you on that wild-goose chase.

So here's a long version of what I found by adding debug output all over this code:

1. `setData` (https://github.com/laravel/framework/blob/6fea6c271086e2ad017a6078e1ee457a2b8c398e/src/Illuminate/Http/JsonResponse.php#L73) gets invoked to handle the response of `Photo::add`.
2. `$data` is our newly added photo model object. So we are, in fact, in your option (1), which you ruled out early on as impossible -- but read on.
3. Our `toJson` gets called (https://github.com/laravel/framework/blob/6fea6c271086e2ad017a6078e1ee457a2b8c398e/src/Illuminate/Http/JsonResponse.php#L78).
4. Our `toJson` calls our `jsonSerialize:` https://github.com/LycheeOrg/Lychee/blob/324188c65316d49e4b2fcd79366922251c1a08cd/app/Models/Extensions/ThrowsConsistentExceptions.php#L134
5. Our `jsonSerialize` calls our `toArray:` https://github.com/LycheeOrg/Lychee/blob/324188c65316d49e4b2fcd79366922251c1a08cd/app/Models/Extensions/ThrowsConsistentExceptions.php#L114
6. Which one of these calls fails and throws an exception? The answer is: **none** of them. They all succeed and result in a perfectly formed JSON string.
7. However, back in `setData`, we get to the invocation of `json_last_error` (https://github.com/laravel/framework/blob/6fea6c271086e2ad017a6078e1ee457a2b8c398e/src/Illuminate/Http/JsonResponse.php#L87) which returns `4` (`JSON_ERROR_SYNTAX`)!
8. Yeah, that didn't make any sense to me either, until I added a debug output of `json_last_error` as the very *first* line of `setData` and there it was again: `4`. *It's a stale error that doesn't get cleared in spite of the subsequent json calls being successful*.
9. I thought it was a bug in PHP until I read word-for-word what [the manual](https://www.php.net/manual/en/function.json-last-error.php) says: *Returns the last error (if any) occurred during the last JSON encoding/decoding, which did not specify JSON_THROW_ON_ERROR.* The way I read it, it's _either_ exceptions _or_ `json_last_error`; we can't freely mix and match both and expect things to just keep working. Our `toJson` invokes `json_encode` with `JSON_THROW_ON_ERROR` which apparently not only prevents `json_last_error` from getting set on error but even (crucially for us) from getting cleared on success.
10. I don't know what code stores an error condition in `json_last_error` in the first place and why it does it for me and not for you. I'll see if I can find it but while that may help you reproduce the error in your setup, I don't think that's critical for coming up with a fix.

Reply:
So I verified that on call to `Photo:add`, here: https://github.com/LycheeOrg/Lychee/blob/324188c65316d49e4b2fcd79366922251c1a08cd/app/Http/Middleware/ContentType.php#L49 `json_last_error` is still `0`, but here: https://github.com/LycheeOrg/Lychee/blob/324188c65316d49e4b2fcd79366922251c1a08cd/app/Http/Requests/BaseApiRequest.php#L79 it's already `4`.
There are some 15 stack frames of Laravel in-between these two so pinpointing the culprit without a debugger is a major PITA. But if you have suggestions how to narrow it down, I'm willing to experiment...

Reply:
> Now, I owe you an apology for giving you an abbreviated version of my findings in my previous message, which sent you on that wild-goose chase.

No problem. I really appreciate that some takes the time to thoroughly test my code and look into it. My apologies that I cause so much trouble. You are really talented at finding weird bugs. :D I like that.

> 9. I thought it was a bug in PHP until I read word-for-word what [the manual](https://www.php.net/manual/en/function.json-last-error.php) says: *Returns the last error (if any) occurred during the last JSON encoding/decoding, which did not specify JSON_THROW_ON_ERROR.*

Yes, I only read this _after_ my long, previously reply, too. Now, everything makes perfect sense to me.

> The way I read it, it's _either_ exceptions _or_ `json_last_error`; we can't freely mix and match both and expect things to just keep working.

Here, I would like to object. You can mix it, but one must only call `json_last_error`, if one has called one of the json methods before without `JSON_THROW_ON_ERROR`. IMHO, the bug is actually in [`JsonResponse::setData`](https://github.com/laravel/framework/blob/6fea6c271086e2ad017a6078e1ee457a2b8c398e/src/Illuminate/Http/JsonResponse.php#L73). This method calls `json_last_error` although it cannot guarantee, that `toJson` has called `json_encode` at all. Just think of a trivial object which implements `Jsonable` but realizes the encoding in a completely different way, e.g. simply returns a valid JSON string. This way a previous, stalled Json Error would not be cleared either. You said, that `JsonResponse::setData` is written in a way which assumes that `toJson` does not throw an exception. I would like to rephrase that: `JsonResponse::setData` is written in a way which assumes that `toJson` uses `json_encode`. And that's a bug, IMHO.

> 10. I don't know what code stores an error condition in `json_last_error` in the first place and why it does it for me and not for you. I'll see if I can find it but while that may help you reproduce the error in your setup, I don't think that's critical for coming up with a fix.

Right. This sounds like another bug in Laravel. As our HTTP request at hand is not even a Json request, but uses Form Data, I guess that something in the middleware calls `json_decode` and does not accurately handle the error condition. Probably something which reads a configuration from a JSON-encoded file. This could also be an explanation why it fails for you but not for me.

Just to pin point the cause further. Could you once try to repreduce the problem with all development code being disabled? I know that the Laravel debug bar, Clockwork and so on, inject a lot JSON for communication between the backend and the frontend. I would like to rule out that the problem is there. Could you run a `composer install --no-dev`, and set `ENVIRONMENT` to `production` as well as `DEBUG_BAR| to `false` in your `.env` file?

Now, since I know for what I have to look out, I am certain I can use my debugger to trace for any call to `json_decode`. I am certain I will find the problematic code.

IMHO, there are three things to do:

1. Issue to bug report to Laravel for the problematic `setData` method. Justification see above.
2. Issue a bug report to Laravel for the actual root cause of this issue after we have found it.
3. Fix our `toJson` method so that it does not trigger the problematic code in `toData`. I already have an idea how to do it.

Reply:
> You are really talented at finding weird bugs. :D

I've heard that before :wink:.

> You said, that `JsonResponse::setData` is written in a way which assumes that `toJson` does not throw an exception. I would like to rephrase that: `JsonResponse::setData` is written in a way which assumes that `toJson` uses `json_encode`. And that's a bug, IMHO.

Agreed.

> Could you once try to repreduce the problem with all development code being disabled? I know that the Laravel debug bar, Clockwork and so on, inject a lot JSON for communication between the backend and the frontend. I would like to rule out that the problem is there. Could you run a `composer install --no-dev`, and set `ENVIRONMENT` to `production` as well as `DEBUG_BAR| to `false`in your`.env` file?

I tried. It still fails in the same way. The message in the front end is different (`Server error` rather than `Syntax error`, I think) but that's to be expected with debugging disabled.

> IMHO, there are three things to do:
> 
>     1. Issue to bug report to Laravel for the problematic `setData` method. Justification see above.

Given your past encounters with them, I'm not overly optimistic about the outcome, but it would be a shame not to try...

>     2. Issue a bug report to Laravel for the actual root cause of this issue after we have found it.

I wish I knew some easy way to intercept all calls to `json_decode`. I know how to do it in C for dynamically linked code but no idea about PHP...

>     3. Fix our `toJson` method so that it does not trigger the problematic code in `toData`. I already have an idea how to do it.

I think an easy way would be to add `json_encode([]);` as the first line of our `toJson` to clear any old errors...

Reply:
> >
> >  1. Issue to bug report to Laravel for the problematic `setData` method. Justification see above.
> >
> 
> Given your past encounters with them, I'm not overly optimistic about the outcome, but it would be a shame not to try...
> 

Exactly what I thought.

> >
> >  2. Issue a bug report to Laravel for the actual root cause of this issue after we have found it.
> > 
> 
> I wish I knew some easy way to intercept all calls to `json_decode`.
>

You don't know the power of [Xdebug](https://xdebug.org/) and a proper IDE. :stuck_out_tongue_closed_eyes: It took me 5min to trace all function calls to `json_encode` and properly will take 20min to write it down pleasantly.

As you mentioned that the bug also occurs in productive mode, here are the methods which call `json_encode` regularly:

 - [`\Illuminate\Encryption\Encrypter::getJsonPayload()`, line 215](https://github.com/laravel/framework/blob/6fea6c271086e2ad017a6078e1ee457a2b8c398e/src/Illuminate/Encryption/Encrypter.php#L213-L228)
   <details>
   <summary>Backtrace</summary>
  
       Encrypter.php:215, json_decode()
       Encrypter.php:215, Illuminate\Encryption\Encrypter->getJsonPayload()
       Encrypter.php:158, Illuminate\Encryption\Encrypter->decrypt()
       EncryptCookies.php:110, Illuminate\Cookie\Middleware\EncryptCookies->decryptCookie()
       EncryptCookies.php:84, Illuminate\Cookie\Middleware\EncryptCookies->decrypt()
       EncryptCookies.php:67, Illuminate\Cookie\Middleware\EncryptCookies->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       AcceptContentType.php:61, App\Http\Middleware\AcceptContentType->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       Pipeline.php:103, Illuminate\Pipeline\Pipeline->then()
       Router.php:723, Illuminate\Routing\Router->runRouteWithinStack()
       Router.php:698, Illuminate\Routing\Router->runRoute()
       Router.php:662, Illuminate\Routing\Router->dispatchToRoute()
       Router.php:651, Illuminate\Routing\Router->dispatch()
       Kernel.php:167, Illuminate\Foundation\Http\Kernel->Illuminate\Foundation\Http\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:164-168}()
       Pipeline.php:128, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:126-132}()
       DisableBrowserCache.php:19, Livewire\DisableBrowserCache->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       SecureHeadersMiddleware.php:21, Bepsvpt\SecureHeaders\SecureHeadersMiddleware->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       TransformsRequest.php:21, Illuminate\Foundation\Http\Middleware\TransformsRequest->handle()
       ConvertEmptyStringsToNull.php:31, Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       TransformsRequest.php:21, Illuminate\Foundation\Http\Middleware\TransformsRequest->handle()
       TrimStrings.php:40, Illuminate\Foundation\Http\Middleware\TrimStrings->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       ValidatePostSize.php:27, Illuminate\Foundation\Http\Middleware\ValidatePostSize->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       PreventRequestsDuringMaintenance.php:86, Illuminate\Foundation\Http\Middleware\PreventRequestsDuringMaintenance->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       TrustProxies.php:57, Fideloper\Proxy\TrustProxies->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       FixStatusCode.php:31, App\Http\Middleware\FixStatusCode->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       Pipeline.php:103, Illuminate\Pipeline\Pipeline->then()
       Kernel.php:142, Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter()
       Kernel.php:111, Illuminate\Foundation\Http\Kernel->handle()
       index.php:74, {main}()
   </details>

 - [`\Illuminate\Http\Request::json()`, line 384](https://github.com/laravel/framework/blob/6fea6c271086e2ad017a6078e1ee457a2b8c398e/src/Illuminate/Http/Request.php#L381-L392)
   <details>
   <summary>Backtrace</summary>

       Request.php:384, json_decode()
       Request.php:384, Illuminate\Http\Request->json()
       Request.php:435, Illuminate\Http\Request::createFrom()
       FormRequestServiceProvider.php:34, Illuminate\Foundation\Providers\FormRequestServiceProvider->Illuminate\Foundation\Providers\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Foundation/Providers/FormRequestServiceProvider.php:33-37}()
       Container.php:1265, Illuminate\Container\Container->fireCallbackArray()
       Container.php:1212, Illuminate\Container\Container->fireResolvingCallbacks()
       Container.php:778, Illuminate\Container\Container->resolve()
       Application.php:851, Illuminate\Foundation\Application->resolve()
       Container.php:694, Illuminate\Container\Container->make()
       Application.php:836, Illuminate\Foundation\Application->make()
       RouteDependencyResolverTrait.php:79, Illuminate\Routing\ControllerDispatcher->transformDependency()
       RouteDependencyResolverTrait.php:48, Illuminate\Routing\ControllerDispatcher->resolveMethodDependencies()
       RouteDependencyResolverTrait.php:28, Illuminate\Routing\ControllerDispatcher->resolveClassMethodDependencies()
       ControllerDispatcher.php:41, Illuminate\Routing\ControllerDispatcher->dispatch()
       Route.php:262, Illuminate\Routing\Route->runController()
       Route.php:205, Illuminate\Routing\Route->run()
       Router.php:721, Illuminate\Routing\Router->Illuminate\Routing\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Routing/Router.php:719-723}()
       Pipeline.php:128, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:126-132}()
       ContentType.php:49, App\Http\Middleware\ContentType->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       SubstituteBindings.php:50, Illuminate\Routing\Middleware\SubstituteBindings->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       VerifyCsrfToken.php:78, Illuminate\Foundation\Http\Middleware\VerifyCsrfToken->handle()
       VerifyCsrfToken.php:61, App\Http\Middleware\VerifyCsrfToken->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       AuthenticateSession.php:39, Illuminate\Session\Middleware\AuthenticateSession->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       ShareErrorsFromSession.php:49, Illuminate\View\Middleware\ShareErrorsFromSession->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       StartSession.php:121, Illuminate\Session\Middleware\StartSession->handleStatefulRequest()
       StartSession.php:64, Illuminate\Session\Middleware\StartSession->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       AddQueuedCookiesToResponse.php:37, Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       EncryptCookies.php:67, Illuminate\Cookie\Middleware\EncryptCookies->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       AcceptContentType.php:61, App\Http\Middleware\AcceptContentType->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       Pipeline.php:103, Illuminate\Pipeline\Pipeline->then()
       Router.php:723, Illuminate\Routing\Router->runRouteWithinStack()
       Router.php:698, Illuminate\Routing\Router->runRoute()
       Router.php:662, Illuminate\Routing\Router->dispatchToRoute()
       Router.php:651, Illuminate\Routing\Router->dispatch()
       Kernel.php:167, Illuminate\Foundation\Http\Kernel->Illuminate\Foundation\Http\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:164-168}()
       Pipeline.php:128, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:126-132}()
       DisableBrowserCache.php:19, Livewire\DisableBrowserCache->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       SecureHeadersMiddleware.php:21, Bepsvpt\SecureHeaders\SecureHeadersMiddleware->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       TransformsRequest.php:21, Illuminate\Foundation\Http\Middleware\TransformsRequest->handle()
       ConvertEmptyStringsToNull.php:31, Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       TransformsRequest.php:21, Illuminate\Foundation\Http\Middleware\TransformsRequest->handle()
       TrimStrings.php:40, Illuminate\Foundation\Http\Middleware\TrimStrings->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       ValidatePostSize.php:27, Illuminate\Foundation\Http\Middleware\ValidatePostSize->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       PreventRequestsDuringMaintenance.php:86, Illuminate\Foundation\Http\Middleware\PreventRequestsDuringMaintenance->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       TrustProxies.php:57, Fideloper\Proxy\TrustProxies->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       FixStatusCode.php:31, App\Http\Middleware\FixStatusCode->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       Pipeline.php:103, Illuminate\Pipeline\Pipeline->then()
       Kernel.php:142, Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter()
       Kernel.php:111, Illuminate\Foundation\Http\Kernel->handle()
       index.php:74, {main}()
   </details>

 - [`\PHPExif\Adapter\Exiftool::getExifFromFile()`, line 151](https://github.com/LycheeOrg/php-exif/blob/a4e6f5c5bf30f1e240ec9a68b4433458c71c5d30/lib/PHPExif/Adapter/Exiftool.php#L128-L172)
   <details>
   <summary>Backtrace</summary>

       Exiftool.php:151, json_decode()
       Exiftool.php:151, PHPExif\Adapter\Exiftool->getExifFromFile()
       Reader.php:108, PHPExif\Reader\Reader->read()
       Extractor.php:162, App\Metadata\Extractor->extract()
       Create.php:136, App\Actions\Photo\Create->loadFileMetadata()
       Create.php:76, App\Actions\Photo\Create->add()
       PhotoController.php:103, App\Http\Controllers\PhotoController->add()
       Controller.php:54, Illuminate\Routing\Controller->callAction()
       ControllerDispatcher.php:45, Illuminate\Routing\ControllerDispatcher->dispatch()
       Route.php:262, Illuminate\Routing\Route->runController()
       Route.php:205, Illuminate\Routing\Route->run()
       Router.php:721, Illuminate\Routing\Router->Illuminate\Routing\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Routing/Router.php:719-723}()
       Pipeline.php:128, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:126-132}()
       ContentType.php:49, App\Http\Middleware\ContentType->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       SubstituteBindings.php:50, Illuminate\Routing\Middleware\SubstituteBindings->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       VerifyCsrfToken.php:78, Illuminate\Foundation\Http\Middleware\VerifyCsrfToken->handle()
       VerifyCsrfToken.php:61, App\Http\Middleware\VerifyCsrfToken->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       AuthenticateSession.php:39, Illuminate\Session\Middleware\AuthenticateSession->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       ShareErrorsFromSession.php:49, Illuminate\View\Middleware\ShareErrorsFromSession->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       StartSession.php:121, Illuminate\Session\Middleware\StartSession->handleStatefulRequest()
       StartSession.php:64, Illuminate\Session\Middleware\StartSession->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       AddQueuedCookiesToResponse.php:37, Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       EncryptCookies.php:67, Illuminate\Cookie\Middleware\EncryptCookies->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       AcceptContentType.php:61, App\Http\Middleware\AcceptContentType->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       Pipeline.php:103, Illuminate\Pipeline\Pipeline->then()
       Router.php:723, Illuminate\Routing\Router->runRouteWithinStack()
       Router.php:698, Illuminate\Routing\Router->runRoute()
       Router.php:662, Illuminate\Routing\Router->dispatchToRoute()
       Router.php:651, Illuminate\Routing\Router->dispatch()
       Kernel.php:167, Illuminate\Foundation\Http\Kernel->Illuminate\Foundation\Http\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:164-168}()
       Pipeline.php:128, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:126-132}()
       DisableBrowserCache.php:19, Livewire\DisableBrowserCache->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       SecureHeadersMiddleware.php:21, Bepsvpt\SecureHeaders\SecureHeadersMiddleware->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       TransformsRequest.php:21, Illuminate\Foundation\Http\Middleware\TransformsRequest->handle()
       ConvertEmptyStringsToNull.php:31, Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       TransformsRequest.php:21, Illuminate\Foundation\Http\Middleware\TransformsRequest->handle()
       TrimStrings.php:40, Illuminate\Foundation\Http\Middleware\TrimStrings->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       ValidatePostSize.php:27, Illuminate\Foundation\Http\Middleware\ValidatePostSize->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       PreventRequestsDuringMaintenance.php:86, Illuminate\Foundation\Http\Middleware\PreventRequestsDuringMaintenance->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       TrustProxies.php:57, Fideloper\Proxy\TrustProxies->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       FixStatusCode.php:31, App\Http\Middleware\FixStatusCode->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       Pipeline.php:103, Illuminate\Pipeline\Pipeline->then()
       Kernel.php:142, Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter()
       Kernel.php:111, Illuminate\Foundation\Http\Kernel->handle()
       index.php:74, {main}()
   </details>

And here are the additional calls to `json_decode`, it debugging tools are enabled

 - [`Clockwork\DataSource\PhpDataSource::getRequestData()`, line 55](https://github.com/itsgoingd/clockwork/blob/9df41432da1d8cb39c7fda383ddcc02231c83ff3/Clockwork/DataSource/PhpDataSource.php#L49-L60)
   <details>
   <summary>Backtrace</summary>

       PhpDataSource.php:55, json_decode()
       PhpDataSource.php:55, Clockwork\DataSource\PhpDataSource->getRequestData()
       PhpDataSource.php:20, Clockwork\DataSource\PhpDataSource->resolve()
       Clockwork.php:59, Clockwork\Clockwork->resolveRequest()
       ClockworkSupport.php:382, Clockwork\Support\Laravel\ClockworkSupport->processRequest()
       ClockworkMiddleware.php:30, Clockwork\Support\Laravel\ClockworkMiddleware->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       Pipeline.php:103, Illuminate\Pipeline\Pipeline->then()
       Kernel.php:142, Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter()
       Kernel.php:111, Illuminate\Foundation\Http\Kernel->handle()
       index.php:74, {main}()
   </details>

 - [`\Barryvdh\Debugbar\Storage\FilesystemStorage::get()`, line 84](https://github.com/barryvdh/laravel-debugbar/blob/b2d39072d094ccd060e27125352985d77aa60dc0/src/Storage/FilesystemStorage.php#L82-L85)
   <details>
   <summary>Backtrace</summary>

       FilesystemStorage.php:84, json_decode()
       FilesystemStorage.php:84, Barryvdh\Debugbar\Storage\FilesystemStorage->get()
       OpenHandler.php:106, DebugBar\OpenHandler->get()
       OpenHandler.php:61, call_user_func:{/var/www/localhost/lychee/vendor/maximebf/debugbar/src/DebugBar/OpenHandler.php:61}()
       OpenHandler.php:61, DebugBar\OpenHandler->handle()
       OpenHandlerController.php:15, Barryvdh\Debugbar\Controllers\OpenHandlerController->handle()
       Controller.php:54, Illuminate\Routing\Controller->callAction()
       ControllerDispatcher.php:45, Illuminate\Routing\ControllerDispatcher->dispatch()
       Route.php:262, Illuminate\Routing\Route->runController()
       Route.php:205, Illuminate\Routing\Route->run()
       Router.php:721, Illuminate\Routing\Router->Illuminate\Routing\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Routing/Router.php:719-723}()
       Pipeline.php:128, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:126-132}()
       BaseController.php:29, Barryvdh\Debugbar\Controllers\BaseController->Barryvdh\Debugbar\Controllers\{closure:/var/www/localhost/lychee/vendor/barryvdh/laravel-debugbar/src/Controllers/BaseController.php:25-30}()
       Pipeline.php:149, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       DebugbarEnabled.php:41, Barryvdh\Debugbar\Middleware\DebugbarEnabled->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       Pipeline.php:103, Illuminate\Pipeline\Pipeline->then()
       Router.php:723, Illuminate\Routing\Router->runRouteWithinStack()
       Router.php:698, Illuminate\Routing\Router->runRoute()
       Router.php:662, Illuminate\Routing\Router->dispatchToRoute()
       Router.php:651, Illuminate\Routing\Router->dispatch()
       Kernel.php:167, Illuminate\Foundation\Http\Kernel->Illuminate\Foundation\Http\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:164-168}()
       Pipeline.php:128, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:126-132}()
       DisableBrowserCache.php:19, Livewire\DisableBrowserCache->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       InjectDebugbar.php:67, Barryvdh\Debugbar\Middleware\InjectDebugbar->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       SecureHeadersMiddleware.php:21, Bepsvpt\SecureHeaders\SecureHeadersMiddleware->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       TransformsRequest.php:21, Illuminate\Foundation\Http\Middleware\TransformsRequest->handle()
       ConvertEmptyStringsToNull.php:31, Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       TransformsRequest.php:21, Illuminate\Foundation\Http\Middleware\TransformsRequest->handle()
       TrimStrings.php:40, Illuminate\Foundation\Http\Middleware\TrimStrings->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       ValidatePostSize.php:27, Illuminate\Foundation\Http\Middleware\ValidatePostSize->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       PreventRequestsDuringMaintenance.php:86, Illuminate\Foundation\Http\Middleware\PreventRequestsDuringMaintenance->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       TrustProxies.php:57, Fideloper\Proxy\TrustProxies->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       FixStatusCode.php:31, App\Http\Middleware\FixStatusCode->handle()
       Pipeline.php:167, Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure:/var/www/localhost/lychee/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:143-174}()
       Pipeline.php:103, Illuminate\Pipeline\Pipeline->then()
       Kernel.php:142, Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter()
       Kernel.php:111, Illuminate\Foundation\Http\Kernel->handle()
       index.php:74, {main}()
   </details>

So, there are only five options in total where it can happen! :tada:

> 
> >
> >  3. Fix our `toJson` method so that it does not trigger the problematic code in `toData`. I already have an idea how to do it.
> >
> 
> I think an easy way would be to add `json_encode([]);` as the first line of our `toJson` to clear any old errors...

Well, I had something else in mind. See here: https://github.com/LycheeOrg/Lychee/pull/1217/commits/33baac2c7a1cd08607fbbc02d77317a009eceb73

Reply:
This is the one:

[`\Illuminate\Http\Request::json()`, line 384](https://github.com/laravel/framework/blob/6fea6c271086e2ad017a6078e1ee457a2b8c398e/src/Illuminate/Http/Request.php#L381-L392)

```
            $this->json = new ParameterBag((array) json_decode($this->getContent(), true));
```
It's invoked just once while handling `Photo:add` and `$this->getContent()` is an empty string. Hence the parse error.

[`\PHPExif\Adapter\Exiftool::getExifFromFile()`, line 151](https://github.com/LycheeOrg/php-exif/blob/a4e6f5c5bf30f1e240ec9a68b4433458c71c5d30/lib/PHPExif/Adapter/Exiftool.php#L128-L172)

This one is not relevant when `has_exiftool = 0`. For me, when it runs, it actually masks the problem, by clearing `json_last_error`.

> > I think an easy way would be to add `json_encode([]);` as the first line of our `toJson` to clear any old errors...
> 
> Well, I had something else in mind. See here: [33baac2](https://github.com/LycheeOrg/Lychee/commit/33baac2c7a1cd08607fbbc02d77317a009eceb73)

Yes, that's better -- no extra calls just to clear the stored error code. I confirmed that it works.

Reply:
> This is the one:
> 
> [`\Illuminate\Http\Request::json()`, line 384](https://github.com/laravel/framework/blob/6fea6c271086e2ad017a6078e1ee457a2b8c398e/src/Illuminate/Http/Request.php#L381-L392)
> 
> ```
>             $this->json = new ParameterBag((array) json_decode($this->getContent(), true));
> ```
> It's invoked just once while handling `Photo:add` and `$this->getContent()` is an empty string. Hence the parse error.
> 

Strangely, on my machine it is called several times and `getContent` returns `null` (not the empty string) and this is properly decoded as `null` again. Why these differences? I smell another Laravel bug here.

Although the original problem is now fixed, I would like to ask you one more time to determine if the method is called before or after our controller method is hit, i.e. does it happen on the way in or out, provide a backtrace and maybe a dump of the other variables in the scope.

I would like to file a Laravel bug on this, too, if we have enough information for that.

Reply:
> Although the original problem is now fixed, I would like to ask you one more time to determine if the method is called before or after our controller method is hit, i.e. does it happen on the way in or out, provide a backtrace and maybe a dump of the other variables in the scope.

It happens on the way in. I believe the backtrace is essentially the same as yours, but here it is again:
```
#0  Illuminate\Http\Request->json() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Http/Request.php:448]
#1  Illuminate\Http\Request::createFrom() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Providers/FormRequestServiceProvider.php:37]
#2  Illuminate\Foundation\Providers\FormRequestServiceProvider->Illuminate\Foundation\Providers\{closure}() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Container/Container.php:1270]
#3  Illuminate\Container\Container->fireCallbackArray() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Container/Container.php:1216]
#4  Illuminate\Container\Container->fireResolvingCallbacks() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Container/Container.php:782]
#5  Illuminate\Container\Container->resolve() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Application.php:857]
#6  Illuminate\Foundation\Application->resolve() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Container/Container.php:697]
#7  Illuminate\Container\Container->make() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Application.php:842]
#8  Illuminate\Foundation\Application->make() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/RouteDependencyResolverTrait.php:84]
#9  Illuminate\Routing\ControllerDispatcher->transformDependency() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/RouteDependencyResolverTrait.php:51]
#10 Illuminate\Routing\ControllerDispatcher->resolveMethodDependencies() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/RouteDependencyResolverTrait.php:31]
#11 Illuminate\Routing\ControllerDispatcher->resolveClassMethodDependencies() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php:41]
#12 Illuminate\Routing\ControllerDispatcher->dispatch() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/Route.php:262]
#13 Illuminate\Routing\Route->runController() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/Route.php:205]
#14 Illuminate\Routing\Route->run() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/Router.php:721]
#15 Illuminate\Routing\Router->Illuminate\Routing\{closure}() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:128]
#16 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/app/Http/Middleware/ContentType.php:49]
#17 App\Http\Middleware\ContentType->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#18 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php:50]
#19 Illuminate\Routing\Middleware\SubstituteBindings->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#20 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php:78]
#21 Illuminate\Foundation\Http\Middleware\VerifyCsrfToken->handle() called at [Lychee-json/app/Http/Middleware/VerifyCsrfToken.php:61]
#22 App\Http\Middleware\VerifyCsrfToken->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#23 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Session/Middleware/AuthenticateSession.php:39]
#24 Illuminate\Session\Middleware\AuthenticateSession->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#25 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php:49]
#26 Illuminate\View\Middleware\ShareErrorsFromSession->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#27 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php:121]
#28 Illuminate\Session\Middleware\StartSession->handleStatefulRequest() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php:64]
#29 Illuminate\Session\Middleware\StartSession->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#30 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php:37]
#31 Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#32 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php:67]
#33 Illuminate\Cookie\Middleware\EncryptCookies->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#34 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/app/Http/Middleware/AcceptContentType.php:61]
#35 App\Http\Middleware\AcceptContentType->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#36 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:103]
#37 Illuminate\Pipeline\Pipeline->then() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/Router.php:723]
#38 Illuminate\Routing\Router->runRouteWithinStack() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/Router.php:698]
#39 Illuminate\Routing\Router->runRoute() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/Router.php:662]
#40 Illuminate\Routing\Router->dispatchToRoute() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Routing/Router.php:651]
#41 Illuminate\Routing\Router->dispatch() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:167]
#42 Illuminate\Foundation\Http\Kernel->Illuminate\Foundation\Http\{closure}() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:128]
#43 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/livewire/livewire/src/DisableBrowserCache.php:19]
#44 Livewire\DisableBrowserCache->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#45 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/barryvdh/laravel-debugbar/src/Middleware/InjectDebugbar.php:67]
#46 Barryvdh\Debugbar\Middleware\InjectDebugbar->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#47 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/bepsvpt/secure-headers/src/SecureHeadersMiddleware.php:21]
#48 Bepsvpt\SecureHeaders\SecureHeadersMiddleware->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#49 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php:21]
#50 Illuminate\Foundation\Http\Middleware\TransformsRequest->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ConvertEmptyStringsToNull.php:31]
#51 Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#52 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php:21]
#53 Illuminate\Foundation\Http\Middleware\TransformsRequest->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TrimStrings.php:40]
#54 Illuminate\Foundation\Http\Middleware\TrimStrings->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#55 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php:27]
#56 Illuminate\Foundation\Http\Middleware\ValidatePostSize->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#57 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php:86]
#58 Illuminate\Foundation\Http\Middleware\PreventRequestsDuringMaintenance->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#59 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/fideloper/proxy/src/TrustProxies.php:57]
#60 Fideloper\Proxy\TrustProxies->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#61 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/app/Http/Middleware/FixStatusCode.php:31]
#62 App\Http\Middleware\FixStatusCode->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#63 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/itsgoingd/clockwork/Clockwork/Support/Laravel/ClockworkMiddleware.php:24]
#64 Clockwork\Support\Laravel\ClockworkMiddleware->handle() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167]
#65 Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:103]
#66 Illuminate\Pipeline\Pipeline->then() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:142]
#67 Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter() called at [Lychee-json/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:111]
#68 Illuminate\Foundation\Http\Kernel->handle() called at [Lychee-json/public/index.php:74]
```

I'm not sure how best to dump the relevant variables. If I dump the backtrace without passing `DEBUG_BACKTRACE_IGNORE_ARGS`, the output is unmanagable (>700 MB). `var_dump($this->toArray());` doesn't seem helpful. Any suggestions?

Reply:
> I'm not sure how best to dump the relevant variables. If I dump the backtrace without passing `DEBUG_BACKTRACE_IGNORE_ARGS`, the output is unmanagable (>700 MB). `var_dump($this->toArray());` doesn't seem helpful. Any suggestions?

Never mind. I thought it would be nice to find out, why `$content` equals the empty string `''` for you and `null` for me. But this is not worth and I already figured out that the problem is actually worse than I thought. Here are the two Laravel bug reports:

 - https://github.com/laravel/framework/issues/42106
 - https://github.com/laravel/framework/issues/42107

Let's see what happens. Actually, the credits are yours for finding these bugs.